### PR TITLE
Scaffolder gitlab oauth token

### LINKED ISSUE #26596: 🐛 Bug Report: Incorrect token usage in scaffolder action 'gitlab:group:ensureExists'
### 📜 Description

When using the repoUrlPicker to retrieve an oauth token from gitlab, and then trying to use that token to call the `gitlab:group:ensureExists` action, i get a '401 unauthorized' from gitlab.
The reason for this is that for REST api calls using oauth tokens, you're supposed to use the `Authorization` header instead of the `PRIVATE-TOKEN` header used with gitlab personal/project/group access tokens.

The issue is located in the `plugins/scaffolder-backend-module-gitlab/src/actions/gitlabGroupEnsureExists.ts#L68`, which directly creates a Gitlab client, instead of using the `createGitlabApi()` method from the `helper.ts`.

### 👍 Expected behavior

Searching and creating groups should work

### 👎 Actual Behavior with Screenshots

It fails with a '401 unauthorized'

### 👟 Reproduction steps

Create a software template which uses the `RepoUrlPicker` to select a gitlab repository:
```yaml
        repoUrl:
          title: Repository Location
          type: string
          ui:field: RepoUrlPicker
          ui:options:
            requestUserCredentials:
              secretsKey: USER_OAUTH_TOKEN
              additionalScopes:
                - api
                - read_api
                - read_repository
                - write_repository
            allowedHosts:
              - git.yourcompany.com
```

then use the `gitlab:groups:ensureExists` action to ensure a group exists

```yaml
    - id: create-groups
      name: Create Groups
      action: gitlab:group:ensureExists
      input:
        token: ${{ secrets.USER_OAUTH_TOKEN }}
        path: 
          - foo
          - bar
```

### 📃 Provide the context for the Bug.

https://docs.gitlab.com/ee/api/rest/#oauth-20-tokens

### 🖥️ Your Environment

OS:   Darwin 23.6.0 - darwin/arm64
node: v20.16.0
yarn: 1.22.22
cli:  0.27.0 (installed)
backstage:  1.27.3

Dependencies:
  @backstage/app-defaults                                          1.5.10
  @backstage/backend-app-api                                       0.9.3
  @backstage/backend-common                                        0.23.3, 0.24.1
  @backstage/backend-defaults                                      0.4.4
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.17
  @backstage/backend-plugin-api                                    0.6.21, 0.7.0, 0.8.1
  @backstage/backend-tasks                                         0.6.1
  @backstage/backend-test-utils                                    0.5.1
  @backstage/catalog-client                                        1.6.6
  @backstage/catalog-model                                         1.6.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.27.0
  @backstage/config-loader                                         1.9.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.2
  @backstage/core-compat-api                                       0.2.8
  @backstage/core-components                                       0.13.10, 0.14.10
  @backstage/core-plugin-api                                       1.9.3
  @backstage/dev-utils                                             1.0.37
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-plugin-api                                   0.7.0
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.30
  @backstage/integration                                           1.14.0
  @backstage/plugin-api-docs                                       0.11.8
  @backstage/plugin-app-backend                                    0.3.73
  @backstage/plugin-app-node                                       0.1.24
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.5
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.17
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.7
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.7
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.2.1
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.19
  @backstage/plugin-auth-backend-module-github-provider            0.1.21
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.21
  @backstage/plugin-auth-backend-module-google-provider            0.1.21
  @backstage/plugin-auth-backend-module-guest-provider             0.1.10
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.19
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.5
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.17
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.6
  @backstage/plugin-auth-backend-module-okta-provider              0.0.17
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.5
  @backstage/plugin-auth-backend                                   0.22.12
  @backstage/plugin-auth-node                                      0.4.17, 0.5.1
  @backstage/plugin-auth-react                                     0.1.5
  @backstage/plugin-bitbucket-cloud-common                         0.2.22
  @backstage/plugin-catalog-backend-module-gitlab                  0.4.1
  @backstage/plugin-catalog-backend-module-msgraph                 0.6.1
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.22
  @backstage/plugin-catalog-backend                                1.25.2
  @backstage/plugin-catalog-common                                 1.0.26
  @backstage/plugin-catalog-graph                                  0.4.8
  @backstage/plugin-catalog-import                                 0.12.2
  @backstage/plugin-catalog-node                                   1.12.6
  @backstage/plugin-catalog-react                                  1.12.3
  @backstage/plugin-catalog                                        1.22.0
  @backstage/plugin-events-backend-module-gitlab                   0.2.10
  @backstage/plugin-events-backend                                 0.3.11
  @backstage/plugin-events-node                                    0.3.10
  @backstage/plugin-home-react                                     0.1.16
  @backstage/plugin-home                                           0.7.9
  @backstage/plugin-org                                            0.6.28
  @backstage/plugin-permission-backend-module-allow-all-policy     0.1.21
  @backstage/plugin-permission-backend                             0.5.48
  @backstage/plugin-permission-common                              0.7.14, 0.8.1
  @backstage/plugin-permission-node                                0.8.2
  @backstage/plugin-permission-react                               0.4.25
  @backstage/plugin-proxy-backend                                  0.5.5
  @backstage/plugin-scaffolder-backend-module-azure                0.1.16
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.14
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.16
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.14
  @backstage/plugin-scaffolder-backend-module-github               0.4.2
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.6
  @backstage/plugin-scaffolder-backend                             1.24.1
  @backstage/plugin-scaffolder-common                              1.5.5
  @backstage/plugin-scaffolder-node                                0.4.10
  @backstage/plugin-scaffolder-react                               1.11.0
  @backstage/plugin-scaffolder                                     1.24.0
  @backstage/plugin-search-backend-module-catalog                  0.2.1
  @backstage/plugin-search-backend-module-explore                  0.2.1
  @backstage/plugin-search-backend-module-pg                       0.5.34
  @backstage/plugin-search-backend-module-techdocs                 0.2.1
  @backstage/plugin-search-backend-node                            1.3.1
  @backstage/plugin-search-backend                                 1.5.16
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.7.14
  @backstage/plugin-search                                         1.4.15
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-backend                               1.10.12
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.13
  @backstage/plugin-techdocs-node                                  1.12.10
  @backstage/plugin-techdocs-react                                 1.2.7
  @backstage/plugin-techdocs                                       1.10.8
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.11
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.10
  @backstage/theme                                                 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by tlanfer**
## Hey, I just made a Pull Request!

Fixes #26596 by using the existing `getClient` utility method.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by tlanfer** [comment]
Fixing the issue was easy enough.
Unfortunately i am not familiar enough with jest and its mocking abilities to figure out how to actually test if the helper actually uses the correct properties when initializing the Gitlab client. 
I see that other actions also dont test for this. I'm happy to just delete the two test cases. Maybe it should rather be added to the tests for that utility module anyway. 

Let me know what it would take to get this merged, or if you see a different, better approach. I'm happy to make more changes.

**Comment by acierto** [comment]
Hi @tlanfer, you can find somegood test examples in another providers, like `scaffolder-backend-module-github` for example.

Each fix or new functionality has to be covered with tests

**Review by mounicarajput — APPROVED** [review]

**Review by mounicarajput — APPROVED** [review]

**Comment by tlanfer** [comment]
Yay, i made CI happy 👏

I decided to mock the `getClient` method from the helper module and then test if it gets called, once with the oauth token if one is provided, and once without a token if none is provided. 
Again, i am definitely not a typescript dev, so i hope i used jest right. So it definitely needs a review by someone who knows what they're doing.
Besides that, i'm happy with the changes now.

File: plugins/scaffolder-backend-module-gitlab/src/actions/gitlabGroupEnsureExists.ts [hunk 1]
```diff
@@ -82,15 +79,15 @@ export const createGitlabGroupEnsureExistsAction = (options: {
...
             pathElement,
             pathElement,
             parent
               ? {
-                  parent_id: parent.id,
+                  parentId: parent.id,
                 }
               : {},
-          );
+          )) as GroupSchema;
```
Comment (jhaals) [inline plugins/scaffolder-backend-module-gitlab/src/actions/gitlabGroupEnsureExists.ts:90]: Is there a reason for casting this?

File: .changeset/sour-phones-fix.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-scaffolder-backend-module-gitlab': patch
+---
+
+Allow the usage of oauth tokens for the gitlab:group:ensureExists action
```
Comment (jhaals) [inline .changeset/sour-phones-fix.md:5]:
```suggestion
Updated `gitlab:group:ensureExists` action to instead use oauth client.
```

**Review by jhaals — CHANGES_REQUESTED** [review]
Thanks for looking into this.

Some minor tweaks and questions

File: plugins/scaffolder-backend-module-gitlab/src/actions/gitlabGroupEnsureExists.ts (hunk 1 above)
Comment (tlanfer) [reply plugins/scaffolder-backend-module-gitlab/src/actions/gitlabGroupEnsureExists.ts]: The `.create()` method returns an `ExpandedGroupSchema` which extends `GroupSchema` but isnt assisgnable.
I have refactored it to only use the `id` property, no need to keep the full `GroupSchema` around.

**Pull request merged**